From d10e708a83f614ebbfa6ad0f472c50a4923ffad0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 18 Nov 2025 17:17:13 +0100
Subject: [PATCH 16/17] server: Set NULL foreground input when switching to
 desktop window.

When called from the application.

Changing foreground window to the desktop seems like a bad idea, but it
also seem to allow later foreground changes back to application windows,
as many D3D tests suggest.

Currently we translate these calls into requests to activate the root
window, which will usually have no effect, so allowing further changes
should be safe.

When the internal flag is set, we actually change the foreground to the
desktop window, which should later forbid any further changes made from
the application side, as this happens when a non-Wine or foreign window
has been focused.
---
 server/queue.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/server/queue.c b/server/queue.c
index 5e601f0be8c..d1f48ead214 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -3896,11 +3896,10 @@ DECL_HANDLER(set_foreground_window)
     reply->send_msg_old = (reply->previous && desktop->foreground_input != queue->input);
     reply->send_msg_new = FALSE;
 
-    if (is_valid_foreground_window( req->handle ) &&
-        (thread = get_window_thread( req->handle )) &&
-        thread->queue->input->desktop == desktop)
+    if (is_valid_foreground_window( req->handle ) && (thread = get_window_thread( req->handle )) &&
+        (input = thread->queue->input) && input->desktop == desktop)
     {
-        set_foreground_input( desktop, thread->queue->input );
+        set_foreground_input( desktop, req->internal || thread->process != get_top_window_owner( desktop ) ? input : NULL );
         reply->send_msg_new = (desktop->foreground_input != queue->input);
     }
     else set_win32_error( ERROR_INVALID_WINDOW_HANDLE );
-- 
2.51.2

